code/game/objects/items/devices/aicard.dm

```diff
@@ -37,71 +37,81 @@
 		overlays.Cut()
 
 /obj/item/aicard/attack_self(mob/user)
-	ui_interact(user)
+	tgui_interact(user)
 
 
-/obj/item/aicard/ui_interact(mob/user, ui_key = "main", var/datum/nanoui/ui = null, var/force_open = 1, var/datum/topic_state/state = GLOB.inventory_state)
-	ui = SSnanoui.try_update_ui(user, src, ui_key, ui, force_open)
+/obj/item/aicard/tgui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = TRUE, datum/tgui/master_ui = null, datum/tgui_state/state = GLOB.tgui_inventory_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
 	if(!ui)
-		ui = new(user, src, ui_key, "aicard.tmpl", "[name]", 600, 400, state = state)
+		ui = new(user, src, ui_key, "AICard", "[name]", 600, 394,  master_ui, state)
 		ui.open()
-		ui.set_auto_update(1)
 
 
-/obj/item/aicard/ui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.inventory_state)
+/obj/item/aicard/tgui_data(mob/user, ui_key = "main", datum/topic_state/state = GLOB.inventory_state)
 	var/data[0]
 
 	var/mob/living/silicon/ai/AI = locate() in src
 	if(istype(AI))
-		data["has_ai"] = 1
+		data["has_ai"] = TRUE
 		data["name"] = AI.name
-		data["hardware_integrity"] = ((AI.health + 100) / 2)
+		data["integrity"] = ((AI.health + 100) / 2)
 		data["radio"] = !AI.aiRadio.disabledAi
 		data["wireless"] = !AI.control_disabled
 		data["operational"] = AI.stat != DEAD
 		data["flushing"] = flush
 
 		var/laws[0]
-		for(var/datum/ai_law/AL in AI.laws.all_laws())
-			laws[++laws.len] = list("index" = AL.get_index(), "law" = sanitize(AL.law))
+		for(var/datum/ai_law/law in AI.laws.all_laws())
+			if(law in AI.laws.ion_laws) // If we're an ion law, give it an ion index code
+				laws.Add(ionnum() + ". " + law.law)
+			else
+				laws.Add(num2text(law.get_index()) + ". " + law.law)
 		data["laws"] = laws
-		data["has_laws"] = laws.len
+		data["has_laws"] = length(AI.laws.all_laws())
+
+	else
+		data["has_ai"] = FALSE // If this isn't passed to tgui, it won't show there isn't a AI in the card.
 
 	return data
 
 
-/obj/item/aicard/Topic(href, href_list, nowindow, state)
+/obj/item/aicard/tgui_act(action, params)
 	if(..())
-		return 1
+		return
 
 	var/mob/living/silicon/ai/AI = locate() in src
 	if(!istype(AI))
-		return 1
+		return
 
 	var/user = usr
-
-	if(href_list["wipe"])
-		var/confirm = alert("Are you sure you want to wipe this card's memory? This cannot be undone once started.", "Confirm Wipe", "Yes", "No")
-		if(confirm == "Yes" && (CanUseTopic(user, state) == STATUS_INTERACTIVE))
+	switch(action)
+		if("wipe")
+			if(flush) // Don't doublewipe.
+				to_chat(user, "<span class='warning'>You are already wiping this AI!</span>")
+				return
```

Comment:
Minor gripe, but please re-add the safety check here since the button could be misclicked.